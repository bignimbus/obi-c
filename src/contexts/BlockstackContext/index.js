import React, {
  useState,
  useEffect,
  createContext,
} from 'react';
import PropTypes from 'prop-types';
import {
  Person,
  signUserOut,
  loadUserData,
  isUserSignedIn,
  isSignInPending,
  redirectToSignIn,
  handlePendingSignIn,
} from 'blockstack';

const BlockstackContext = createContext();

export const PENDING = 'pending';
export const LOADING = 'loading';
export const AUTHENTICATED = 'authenticated';

export const BlockstackContextProvider = ({ children }) => {
  const [user, setUser] = useState(null);
  const [authState, setAuthState] = useState(PENDING);

  const signIn = () => {
    redirectToSignIn();
  };

  const signOut = () => {
    signUserOut(window.location.origin);
  }

  const verifyAuthStatus = async () => {
    if (isUserSignedIn()) {
      const { profile } = loadUserData();
      setUser(new Person(profile));
      setAuthState(AUTHENTICATED);
    } else if (isSignInPending()) {
      try {
        setAuthState(LOADING);
        const { profile } = await handlePendingSignIn();
        setUser(new Person(profile));
        setAuthState(AUTHENTICATED);
      } catch (e) {
        console.warn(e);
      }
    }
  };

  useEffect(() => {
    if (user) return;
    verifyAuthStatus();
  });

  return (
    <BlockstackContext.Provider
      value={{
        user,
        signIn,
        signOut,
        authState,
      }}
    >
      { children }
    </BlockstackContext.Provider>
  );
};

BlockstackContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default BlockstackContext;
