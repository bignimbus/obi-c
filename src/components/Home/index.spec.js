import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import BlockstackContext from '../../contexts/BlockstackContext';
import { default as Home } from '.';

const renderComponent = (props = {}) => renderer.create(
  <BlockstackContext.Provider
    value={{
      user: null,
      signIn: jest.fn(),
      authState: 'foo',
    }}
  >
    <MemoryRouter>
      <Home {...props} />
    </MemoryRouter>
  </BlockstackContext.Provider>
);

describe('Home', () => {
  describe('#render', () => {
    test('app menu is open', () => {
      const element = renderComponent({ appMenuOpen: true });
      expect(element.toJSON()).toMatchSnapshot();
    });

    test('app menu is closed', () => {
      const element = renderComponent();
      expect(element.toJSON()).toMatchSnapshot();
    });
  });
});
