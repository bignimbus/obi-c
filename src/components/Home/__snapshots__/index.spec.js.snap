// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home #render app menu is closed 1`] = `
<div
  aria-label="Tap anywhere to open menu"
  className="home"
  onClick={[Function]}
  role="button"
  tabIndex="0"
>
  <section
    style={
      Object {
        "display": "none",
      }
    }
  >
    <div
      className="app-menu"
    >
      <nav
        className="app-menu__nav"
      >
        <header
          className="app-menu__header"
        >
          <div>
            <section>
              <button
                disabled={false}
                onClick={[MockFunction]}
              >
                <span>
                  Connect to Blockstack
                </span>
              </button>
            </section>
            <footer>
              <button>
                Learn more
              </button>
            </footer>
          </div>
        </header>
        <section
          className="app-menu__section"
        >
          <ul
            className="app-menu__ul"
          >
            <li
              className="app-menu__li"
            >
              <a
                href="/remind"
                onClick={[Function]}
              >
                Remind me
              </a>
            </li>
            <li
              className="app-menu__li"
            >
              <a
                href="/feed"
                onClick={[Function]}
              >
                Feed obi-c
              </a>
            </li>
            <li
              className="app-menu__li"
            >
              <a
                href="/play"
                onClick={[Function]}
              >
                Play with obi-c
              </a>
            </li>
          </ul>
        </section>
      </nav>
    </div>
  </section>
  <section>
    Home
  </section>
</div>
`;

exports[`Home #render app menu is open 1`] = `
<div
  aria-label="Tap anywhere to open menu"
  className="home"
  onClick={[Function]}
  role="button"
  tabIndex="0"
>
  <section
    style={
      Object {
        "display": "block",
      }
    }
  >
    <div
      className="app-menu"
    >
      <nav
        className="app-menu__nav"
      >
        <header
          className="app-menu__header"
        >
          <div>
            <section>
              <button
                disabled={false}
                onClick={[MockFunction]}
              >
                <span>
                  Connect to Blockstack
                </span>
              </button>
            </section>
            <footer>
              <button>
                Learn more
              </button>
            </footer>
          </div>
        </header>
        <section
          className="app-menu__section"
        >
          <ul
            className="app-menu__ul"
          >
            <li
              className="app-menu__li"
            >
              <a
                href="/remind"
                onClick={[Function]}
              >
                Remind me
              </a>
            </li>
            <li
              className="app-menu__li"
            >
              <a
                href="/feed"
                onClick={[Function]}
              >
                Feed obi-c
              </a>
            </li>
            <li
              className="app-menu__li"
            >
              <a
                href="/play"
                onClick={[Function]}
              >
                Play with obi-c
              </a>
            </li>
          </ul>
        </section>
      </nav>
    </div>
  </section>
  <section>
    Home
  </section>
</div>
`;
